### PR TITLE
32-transform-to-and-from-xliff-2

### PR DESCRIPTION
**Pull request by VladislavsPerkanuks**

### CODE REVIEW

File: pkg/convert/xliff2.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	SrcLang language.Tag `xml:"srcLang,attr"`
+	File    file         `xml:"file"`
+}
+type file struct {
+	Units []unit `xml:"unit"`
+}
+
+type unit struct {
+	ID     string  `xml:"id,attr"`
+	Notes  *[]note `xml:"notes>note"` // Set as pointer to avoid empty <notes></notes> when marshalling.
```
Comment (VladislavsPerkanuks) [inline pkg/convert/xliff2.go:26]: I tried many solutions, but this is the only one that works: avoiding `<notes></notes>` when marshaling `xliff2` struct with no `Notes`.
If there is a better way/ways I am happy to hear about them :)

File: pkg/convert/xliff2.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+			Notes:  notes,
+		})
+	}
+
+	data, err := xml.Marshal(&xlf)
+	if err != nil {
+		return nil, fmt.Errorf("marshal model.Messages to XLIFF 2 formatted XML: %w", err)
+	}
+
+	dataWithHeader := append([]byte(xml.Header), data...) // prepend generic XML header
```
Comment (VladislavsPerkanuks) [inline pkg/convert/xliff2.go:91]: Adding XML Declaration Header is recommended
https://www.w3.org/TR/xml/#sec-prolog-dtd

File: pkg/convert/xliff2_test.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        <note category="description">App title</note>
+      </notes>
+      <segment>
+        <source>Diary</source>
+        <target>Agenda</target>
+      </segment>
+    </unit>
+  </file>
+</xliff>`),
+			want: model.Messages{
```
Comment (DarjaLeonova) [inline pkg/convert/xliff2_test.go:48]: Maybe it's better to call it expected?
expected: model.Messages...

File: pkg/convert/xliff2_test.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+	tests := []struct {
+		name    string
+		wantErr error
+		data    []byte
+		want    model.Messages
+	}{
+		{
+			name: "All OK",
+			data: []byte(`<?xml version="1.0" encoding="UTF-8"?>
```
Comment (DarjaLeonova) [inline pkg/convert/xliff2_test.go:24]: And here maybe input?

File: pkg/convert/xliff2_test.go (hunk 3 above)
Comment (VladislavsPerkanuks) [reply pkg/convert/xliff2_test.go:48]: Maybe you are right, but I guess it should be in a separate PR, as if I change I should change in all files to remain consistency

File: pkg/convert/xliff2.go (hunk 1 above)
Comment (janishorsts) [reply pkg/convert/xliff2.go:26]: try `omitempty`

```suggestion
	Notes  []note `xml:"notes>note,omitempty"`
```

Comment (VladislavsPerkanuks) [reply pkg/convert/xliff2.go:26]:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<xliff xmlns="urn:oasis:names:tc:xliff:document:2.0" version="2.0" srcLang="en">
  <file>
    <unit id="Welcome">
      <notes>
        <note category="description">To welcome a new visitor</note>
      </notes>
      <segment>
        <source>Welcome to our website!</source>
      </segment>
    </unit>
    <unit id="Error">
      <notes>
        <note category="description">To inform the user of an error</note>
      </notes>
      <segment>
        <source>Something went wrong. Please try again later.</source>
      </segment>
    </unit>
    <unit id="Feedback">
      <notes></notes>
      <segment>
        <source>We appreciate your feedback. Thank you for using our service.</source>
      </segment>
    </unit>
  </file>
</xliff>

```

Comment (janishorsts) [reply pkg/convert/xliff2.go:26]: Not possible using field tag https://github.com/golang/go/issues/7233

It is possible by implementing `xml.Marshaler`.

File: pkg/convert/xliff2_test.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+				assert.ErrorContains(t, err, tt.wantErr.Error())
+				return
+			}
+
+			if !assert.NoError(t, err) {
+				return
+			}
+
+			// Matches a substring that starts with > and ends with < with zero or more whitespace in between.
+			re := regexp.MustCompile(`>(\s*)<`)
```
Comment (janishorsts) [inline pkg/convert/xliff2_test.go:190]: You can create a test helper function if you use it in multiple tests.

```go
func equalXml(t *testing.T, a, b []byte) boolean {
  t.Helper()

  // ....
}
```

**Review by janishorsts — APPROVED** [review]

**Pull request merged**